dynamicio/mixins.py

```diff
@@ -866,19 +866,26 @@ def _get_producer(self, server: str, **options: MutableMapping[str, Any]) -> Kaf
         }
         return KafkaProducer(**self.__kafka_config)
 
-    def _send_messages(self, df: pd.DataFrame, topic: str) -> None:
+    def _send_messages(self, df: pd.DataFrame, topic: str, show_warnings: bool = True) -> None:
         logger.info(f"Sending {len(df)} messages to Kafka topic:{topic}.")
 
-        indices = df.index.values  # possibly non-unique indices
-        if len(set(indices)) != len(indices):
-            logger.warning("Messages have non-unique keys. This may cause issues in your downstream logic.")
+        indices = df.index.values
+
+        if show_warnings:
+            if self._df_index_is_used_for_keying_messages():
+                if len(set(indices)) != len(indices):  # index is not unique
+                    logger.warning("Messages have non-unique keys. This may cause issues in your downstream logic.")
 
         messages = df.reset_index(drop=True).to_dict("records")
         for idx, message in zip(indices, messages):
             self.__producer.send(topic, key=self.__key_generator(idx, message), value=self.__document_transformer(message))  # type: ignore
 
         self.__producer.flush()  # type: ignore
 
+    def _df_index_is_used_for_keying_messages(self) -> bool:
+        get_constants = operator.attrgetter("co_code", "co_consts")
+        return get_constants(self.__key_generator.__code__) == get_constants((lambda index, _: index).__code__)  # type: ignore
+
```

Comment:
Could you please explain what you want this to achieve?

Reply:
If the user provides a lambda for the `key_generator` I want to compare that against a lambda that would practically assign the `index` as the message's key. 

Ideal scenario: 

* User provides: `key_gen: lambda idx, _: idx` 
* I compare `key_gen == lambda idx, _: idx`
* If true, then... 

I want a way to compare lambdas. This alone already tells me that the interface is not great. 

Regardless, I have resorted to comparing bytecode and constants instead.

Reply:
This is a very weak check, a user could provide any named args to that lambda... way overengineered for a warning message imo. I suggest:
- by default, use the index of each row as the message key
- clients can provide their own keying function in which case that's their problem OR you can warn them then

Either way, there's no problem in not having unique keys. Think of our use cases - they are keyed on entity ids on purpose and they are not unique.

Reply:
I 'll include a heads up in the API docs.